buildserver/environment_test.go

```diff
@@ -0,0 +1,168 @@
+	type tcase struct {
+		Name           string
+		RootURI        lsp.DocumentURI
+		WantImportPath string
+		WantGoPath     string
+		FS             map[string]string
+	}
+
+	cases := []tcase{
+		{
+			Name:           "glide",
+			RootURI:        "git://github.com/alice/pkg",
+			WantImportPath: "alice.org/pkg",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				"glide.yaml": "package: alice.org/pkg",
+			},
+		},
+		{
+			Name:           "canonical",
+			RootURI:        "git://github.com/alice/pkg",
+			WantImportPath: "alice.org/pkg",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				"doc.go": `package pkg // import "alice.org/pkg"`,
+			},
+		},
+		{
+			Name:           "nested-canonical",
+			RootURI:        "git://github.com/alice/pkg",
+			WantImportPath: "alice.org/pkg",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				"http/doc.go": `package http // import "alice.org/pkg/http"`,
+			},
+		},
+		{
+			Name:           "customer-cmd-path",
+			RootURI:        "git://github.com/alice/code",
+			WantImportPath: "alice.org/code",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				"kode/cmd/alice/alice.go": `package http // import "alice.org/code/kode/cmd/alice"`,
+			},
+		},
+		{
+			Name:           "cfg-too-long",
+			RootURI:        "git://github.com/alice/code",
+			WantImportPath: "good.org/code",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				// Not picked up because it is one level too deep.
+				"a/b/c/d/alice.go": `package http // import "alice.org/code/a/b/c/d"`,
+
+				// But users can manually specify it via the config:
+				".sourcegraph/config.json": `{"go": {"RootImportPath": "good.org/code"}}`,
+			},
+		},
+		{
+			Name:           "cfg-invalid-detection",
+			RootURI:        "git://github.com/alice/code",
+			WantImportPath: "alice.org/code",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				// Pretend this code is *not* their actual code, they just have
+				// someone else's code copied into their project (e.g. very poor
+				// vendoring solution). They do *not* want us to use this
+				// canonical import path.
+				"kode/cmd/alice/alice.go": `package http // import "bad.org/code/kode/cmd/alice"`,
+
+				// They can override it via the config:
+				".sourcegraph/config.json": `{"go": {"RootImportPath": "alice.org/code"}}`,
+			},
+		},
+		{
+			Name:           "fallback",
+			RootURI:        "git://github.com/alice/pkg",
+			WantImportPath: "github.com/alice/pkg",
+			WantGoPath:     gopath,
+			FS: map[string]string{
+				"doc.go": `package pkg`,
+			},
+		},
+
+		{
+			Name:           "monorepo",
+			RootURI:        "git://github.com/alice/monorepo",
+			WantImportPath: "",
+			WantGoPath:     "/workspace/third_party:/workspace/code:/",
+			FS: map[string]string{
+				".vscode/settings.json": fmt.Sprintf(`{
+// this JSON document has comments!
+  "go.gopath": "${workspaceRoot}/third_party%s${workspaceRoot}/code",
+}`, string(os.PathListSeparator)),
```

Comment:
nice :)

Reply:
Heh, I found this when the test failed on Travis on Windows.